### ISSUE TITLE
[project] MicroSD card slot module

### ISSUE THREAD

**Issue by NegrilaRares** [UPB-FILS-MA/questions#12]
# Question

## Question
Trying to get data from a MicroSD card to input into a DAC to get an analog input for a speaker.
My DAC features 12 bit resolution with a SPI interface and i was wondering if you have any crates for reading binary or wav files.

Should I struggle with 12 bit wav files or change to 8?

The DAC ignores the last 4 bits of the second data byte by default, and I was wondering if I could make it so that it ignores the whole second byte.
![image](https://github.com/UPB-FILS-MA/questions/assets/129112753/2d3ed7cf-669e-43ea-aca1-5c24e410274c)

I tried transforming a 16 bit wav into a 16 bit binary to then convert it using python into a 12 bit binary.
I'm not sure how much data is actually lost during the conversion. 

![image](https://github.com/UPB-FILS-MA/questions/assets/129112753/ef3210ea-0a55-4ef8-a7bc-d63d722c6a23)
![image](https://github.com/UPB-FILS-MA/questions/assets/129112753/a62b77ec-357c-4fd7-b159-5fbb7f93c384)

Can I use this for MicroSDs: https://github.com/rust-embedded-community/embedded-sdmmc-rs ?  

And would using rp_pico::hal code like this in embassy be possible: https://github.com/rp-rs/rp-hal-boards/blob/main/boards/rp-pico/examples/pico_spi_sd_card.rs#L180 ?

If you could please give me any sort of advice on how I should move forward with the project or any crates I could use.
Thank you in advance.
## Details
- Link to your project repository: https://github.com/UPB-FILS-MA/project-NegrilaRares

## Material read

MicroSD card slot module USED: https://www.optimusdigital.ro/en/memories/1516-microsd-card-slot-module.html
Digital to analog converter USED: https://www.optimusdigital.ro/en/others/1327-dac-mcp4725-module-with-i2c-interface.html?search_query=DAC&results=62

**Comment by NegrilaRares** [comment]
Tried using this https://github.com/rp-rs/rp-hal-boards/blob/main/boards/rp-pico/examples/pico_spi_sd_card.rs#L180 . 

Keep getting an error when initializing SdCard::new from :  https://github.com/rust-embedded-community/embedded-sdmmc-rs .

![image](https://github.com/UPB-FILS-MA/questions/assets/129112753/c77fa1f1-1ba7-4cc7-83b6-9a7cd15d512d)

![image](https://github.com/UPB-FILS-MA/questions/assets/129112753/acf7b836-9aec-4858-9e1e-e33eea3fcf1d)

The lines are the exact ones from my project in case more context is necessary.

I tried compiling this by itself but got the same error: https://github.com/rp-rs/rp-hal-boards/blob/main/boards/rp-pico/examples/pico_spi_sd_card.rs#L180 .

[error.txt](https://github.com/UPB-FILS-MA/questions/files/15044358/error.txt)



**This was solved by using embedded-sdmmc-rs 0.5.0 instead of the 0.7.0.**

**Comment by NegrilaRares** [comment]
After carefully looking through the crates used by the example pico_spi_sd_card from rp-hal-boards I noticed that they used the 0.5.0 version of embedded-sdmmc-rs while i was using the newest 0.7.0 version.

After switching versions the previous errors did no longer appear but i started getting a symbol multiply defined error regarding _critical_section_1_0_acquire.

![image](https://github.com/UPB-FILS-MA/questions/assets/129112753/6554f525-ad8a-4df3-9780-26b85b87ee0e)


**This was solved my using the example without the use of rp2040-hal, making the delay and spi in embassy instead.
For those who wish to make a delay with embassy do note that you will have to use embedded_hal::delay::DelayNs
alongside embassy_time::Delay. Because the Delay of embassy_time implements traits from embedded_hal.**
